File: backend/app/schemas/vendor_profile.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


# --- Profile Schemas ---


class VendorProfileUpdate(BaseModel):
    bio: str | None = None
    founding_year: int | None = None
    certifications: dict | None = None
    specialization_tags: list[str] | None = None


class VendorProfileResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: UUID
    vendor_id: UUID
    display_photo: str | None = None
    bio: str | None = None
    founding_year: int | None = None
    certifications: dict | None = None
    specialization_tags: list[str] | None = None
    quality_score: str  # Decimal serialized as string
    vendor_tier: str
    tier_changed_at: datetime | None = None
    profile_completeness: int
    created_at: datetime
    updated_at: datetime


class VendorPublicProfileResponse(BaseModel):
    model_config = {"from_attributes": True}
    vendor_id: UUID
    vendor_name: str
    display_photo: str | None = None
    bio: str | None = None
    founding_year: int | None = None
    certifications: dict | None = None
    specialization_tags: list[str] | None = None
    quality_score: str
    vendor_tier: str
    profile_completeness: int
    # Aggregated stats
    total_completed_jobs: int = 0
    avg_rating: float | None = None
    total_ratings: int = 0
    first_time_acceptance_rate: float | None = None
    avg_turnaround_hours: float | None = None
    on_time_delivery_rate: float | None = None
    service_areas: list[dict] = []


class PortfolioItemResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: UUID
    property_type: str
    report_category: str
    city: str
    area: str | None = None
    completed_at: datetime | None = None


class PortfolioResponse(BaseModel):
    items: list[PortfolioItemResponse]
    total: int
    page: int
    page_size: int


# --- Rating Schemas ---


class VendorRatingCreate(BaseModel):
    rating: int = Field(..., ge=1, le=5)
    report_request_id: UUID


class VendorRatingResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: UUID
    lender_user_id: UUID
    vendor_id: UUID
    report_request_id: UUID
    rating: int
    created_at: datetime


class VendorRatingSummary(BaseModel):
    vendor_id: UUID
    avg_rating: float | None = None
    total_ratings: int = 0
    rating_distribution: dict[str, int] = {}  # {"1": 5, "2": 3, ...}


# --- Tier Schemas ---


class TierProgressResponse(BaseModel):
    current_tier: str
    tier_changed_at: datetime | None = None
    quality_score: str
    completed_jobs: int
    avg_rating: float | None = None
    first_time_acceptance_rate: float | None = None
    on_time_delivery_rate: float | None = None
    avg_response_hours: float | None = None
    # Next tier requirements
    next_tier: str | None = None
    next_tier_requirements: dict | None = None


class AdminTierOverride(BaseModel):
    vendor_tier: str
    reason: str | None = None

File: frontend/src/types/vendor-profile.ts
export interface VendorProfile {
  id: string;
  vendor_id: string;
  display_photo: string | null;
  bio: string | null;
  founding_year: number | null;
  certifications: Record<string, string> | null;
  specialization_tags: string[] | null;
  quality_score: string;
  vendor_tier: "NEW" | "VERIFIED" | "TOP_VALUER";
  tier_changed_at: string | null;
  profile_completeness: number;
  created_at: string;
  updated_at: string;
}

export interface VendorPublicProfile {
  vendor_id: string;
  vendor_name: string;
  display_photo: string | null;
  bio: string | null;
  founding_year: number | null;
  certifications: Record<string, string> | null;
  specialization_tags: string[] | null;
  quality_score: string;
  vendor_tier: "NEW" | "VERIFIED" | "TOP_VALUER";
  profile_completeness: number;
  total_completed_jobs: number;
  avg_rating: number | null;
  total_ratings: number;
  first_time_acceptance_rate: number | null;
  avg_turnaround_hours: number | null;
  on_time_delivery_rate: number | null;
  service_areas: ServiceAreaInfo[];
}

export interface ServiceAreaInfo {
  city: string;
  areas: string[] | null;
  service_type: string | null;
}

export interface PortfolioItem {
  id: string;
  property_type: string;
  report_category: string;
  city: string;
  area: string | null;
  completed_at: string | null;
}

export interface PortfolioResponse {
  items: PortfolioItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface VendorRating {
  id: string;
  lender_user_id: string;
  vendor_id: string;
  report_request_id: string;
  rating: number;
  created_at: string;
}

export interface VendorRatingSummary {
  vendor_id: string;
  avg_rating: number | null;
  total_ratings: number;
  rating_distribution: Record<string, number>;
}

export interface TierProgress {
  current_tier: "NEW" | "VERIFIED" | "TOP_VALUER";
  tier_changed_at: string | null;
  quality_score: string;
  completed_jobs: number;
  avg_rating: number | null;
  first_time_acceptance_rate: number | null;
  on_time_delivery_rate: number | null;
  avg_response_hours: number | null;
  next_tier: string | null;
  next_tier_requirements: {
    min_completed_jobs?: number;
    min_quality_score?: number;
    max_response_hours?: number;
    current_completed_jobs?: number;
    current_quality_score?: string;
    current_avg_response_hours?: number | null;
  } | null;
}
